isarn-sketches-java/src/main/java/org/isarnproject/sketches/java/TDigest.java

```diff
@@ -0,0 +1,761 @@
+        int end = nclusters - 1;
+        double mend = M;
+        while ((end - beg) > 1) {
+            int mid = (beg + end) / 2;
+            double mmid = ftSum(mid);
+            if (m >= mmid) {
+                beg = mid;
+                mbeg = mmid;
+            } else {
+                end = mid;
+                mend = mmid;
+            }
+        }
+        return beg;
+    }
+
+    // returns the index of a left mass cover
+    // ftSum(j-1) < m <= ftSum(j)
+    private final int lmcovj(double m) {
+        assert nclusters >= 2;
+        assert (m >= 0.0) && (m <= M);
+        int beg = -1;
+        double mbeg = 0.0;
+        int end = nclusters - 1;
+        double mend = M;
+        while ((end - beg) > 1) {
+            int mid = (beg + end) / 2;
+            double mmid = ftSum(mid);
+            if (m <= mmid) {
+                end = mid;
+                mend = mmid;
+            } else {
+                beg = mid;
+                mbeg = mmid;
+            }
+        }
+        return end;
+    }
+
+    // returns the left index of a right-cover
+    private final int rcovj(double x) {
+        int j = Arrays.binarySearch(cent, 0, nclusters, x);
+        // exact match, return its index:
+        if (j >= 0) return j;
+        // x is not a cluster center, get its insertion index:
+        j = -(j + 1);
+        // x is to left of left-most cluster:
+        if (j == 0) return -1;
+        // return the index to the left of x:
+        return j - 1;
+    }
+
+    // cumulative-sum algorithm for a Fenwick tree
+    private final double ftSum(int j) {
+        j += 1;
+        double s = 0.0;
+        while (j > 0) {
+            s += ftre[j];
+            j -= j & (-j); // dec by least significant nonzero bit of j
+        }
+        return s;
+    }
+
+    // increment algorithm for a Fenwick tree
+    private final void ftInc(int j, double w) {
+        j += 1;
+        while (j <= nclusters) {
+            ftre[j] += w;
+            j += j & (-j); // inc by least significant nonzero bit of j
+        }
+    }
+
+    @Override
+    public boolean equals(Object that) {
+        if (!(that instanceof TDigest)) return false;
+        if (this == that) return true;
+        TDigest rhs = (TDigest)that;
+        if (C != rhs.C) return false;
+        if (maxDiscrete != rhs.maxDiscrete) return false;
+        if (nclusters != rhs.nclusters) return false;
+        if (M != rhs.M) return false;
+        if (!equal(cent, rhs.cent, nclusters)) return false;
+        if (!equal(mass, rhs.mass, nclusters)) return false;
+        // if masses are equal, cumulative ftre had better also be equal
+        return true;
+    }
+
+    // I can't believe java just added this to Arrays in java 9
+    static final boolean equal(double[] lhs, double[] rhs, int n) {
+        for (int j = 0; j < n; ++j) {
+            if (lhs[j] != rhs[j]) return false;
+        }
+        return true;
+    }
+
+    @Override
+    public int hashCode() {
```

Comment:
This method makes me nervous, especially in the `nclusters >= 3` case, but in general hashing these structures seems problematic.  Would suggest a detailed comment with rationale.

Reply:
My goal here was to keep hashing O(1), however to also maximize the probability of capturing differences between two structures. I claim (without having any proof) that the likelihood of two t-digests having the same beginning, midpoint, and ending `(x,mass(x),ftre(x))` triples but still being different elsewhere is very small.

Reply:
Also, the hash was primarily to provide standard equality testing logic, and so even if the hash is less discriminating than I believe, it isn't a showstopper.  However, other hashing policies are definitely possible.

Reply:
(I agree with your assessment that it's unlikely to lead to many spurious conflicts.)